src/real_double.cpp

```diff
@@ -39,4 +40,104 @@ int RealDouble::compare(const Basic &o) const
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::atan(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> acot(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::atan(1/static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> asec(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::acos(1/static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> acsc(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::asin(1/static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> sinh(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::sinh(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> cosh(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::cosh(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> tanh(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::tanh(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> coth(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(1/std::tanh(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> asinh(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::asinh(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> acosh(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::acosh(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> atanh(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::atanh(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> acoth(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::atanh(1/static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> log(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::log(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> gamma(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::tgamma(static_cast<const RealDouble &>(x).i));
+    }
+    virtual RCP<const Basic> abs(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return real_double(std::abs(static_cast<const RealDouble &>(x).i));
+    }
+};
+
+Evaluate& RealDouble::get_eval() const {
+    static EvaluateDouble evaluate_double;
+    return evaluate_double;
```

Comment:
I think this is fine for now. But if somehow later somebody removes the word `static` or complicates the code, then the `Evaluate&` reference can become dangling. One solution to this is described in #445, essentially you return `Ptr<Evaluate>` and use it as a pointer. In Release mode this will be raw pointer, thus no speed penalty. But in Debug mode, this will be a pointer to a `UniquePtr`, thus (after we switch to use `Teuchos::RCP` in Debug mode) this will be 100% checked --- if the pointer becomes dangling, because the `EvaluateDouble` somehow went out of scope (i.e. by removing the `static` or some other way), you get a nice exception.